system-addon/lib/DownloadsManager.jsm

```diff
@@ -0,0 +1,146 @@
+    const downloadedItem = element.download;
+    let description;
+    if (element.fileType) {
+      // If we have a file type: '1.5 MB — PNG'
+      description = `${element.sizeStrings.stateLabel} \u2014 ${element.fileType}`;
+    } else {
+      // If we do not have a file type: '1.5 MB'
+      description = `${element.sizeStrings.stateLabel}`;
+    }
+    return {
+      hostname: new URL(downloadedItem.source.url).hostname,
+      url: downloadedItem.source.url,
+      path: downloadedItem.target.path,
+      title: element.displayName,
+      description,
+      referrer: downloadedItem.source.referrer,
+      date_added: downloadedItem.endTime
+    };
+  }
+
+  init(store) {
+    this._store = store;
+    this._browser = Services.appShell.hiddenDOMWindow;
+    this._downloadData = DownloadsCommon.getData(this._browser.ownerGlobal, true, false, true);
+    this._downloadData.addView(this);
+  }
+
+  onDownloadAdded(download) {
+    const elem = new DownloadElement(download, this._browser);
+    const downloadedItem = elem.download;
+    if (!this._viewableDownloadItems.has(downloadedItem.source.url)) {
+      this._viewableDownloadItems.set(downloadedItem.source.url, elem);
+      this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+    }
+  }
+
+  onDownloadRemoved(download) {
+    if (this._viewableDownloadItems.has(download.source.url)) {
+      this._viewableDownloadItems.delete(download.source.url);
+      this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+    }
+  }
+
+  get downloads() {
+    let results = [];
+    for (const url of this._viewableDownloadItems.keys()) {
+      const elem = this._viewableDownloadItems.get(url);
+      // We only want downloads that are less than 36 hours old
+      if (elem && (elem.downloadedSince < RECENT_DOWNLOAD_THRESHOLD)) {
+        const formattedDownloadForHighlights = this.formatDownload(elem);
+        results.push(formattedDownloadForHighlights);
+      }
+    }
+    return results.sort((a, b) => a.date_added < b.date_added);
+  }
+
+  uninit() {
+    if (this._downloadData) {
+      this._downloadData.removeView(this);
+      this._downloadData = null;
+    }
+  }
+
+  onAction(action) {
+    let downloadsCmd;
+    switch (action.type) {
+      case at.COPY_DOWNLOAD_LINK:
+        downloadsCmd = "copyLocation";
+        break;
+      case at.GO_TO_DOWNLOAD_PAGE:
+        downloadsCmd = "openReferrer";
+        break;
+      case at.REMOVE_DOWNLOAD_FILE:
+        downloadsCmd = "delete";
+        break;
+      case at.SHOW_DOWNLOAD_FILE:
+        downloadsCmd = "show";
+        break;
+      case at.OPEN_DOWNLOAD_FILE:
+        downloadsCmd = "open";
+        break;
+      case at.UNINIT:
+        this.uninit();
+        break;
+    }
+    // Call the appropriate downloads command function based on the event we received
+    if (downloadsCmd) {
+      let elem = this._viewableDownloadItems.get(action.data.url);
+      if (elem) {
+        elem[`downloadsCmd_${downloadsCmd}`]();
```

Comment:
I've learned that dynamically building keys for strings, methods, etc. is nice looking but makes it hard to find, e.g., from searchfox. "Are there any more callers of downloadsCmd_delete? No? Baleeted!" Sorry for giving the example with string building in the previous comment. :(

Reply:
😂 no worries!